pkg/sac/allow_fixed_scopes_checker_core.go

```diff
@@ -3,121 +3,288 @@ package sac
+	// Global access granted
+	if c.accessKeys.Cardinality() == 0 &&
+		c.resourceKeys.Cardinality() == 0 &&
+		c.clusterKeys.Cardinality() == 0 &&
+		c.resourceKeys.Cardinality() == 0 {
 		return effectiveaccessscope.UnrestrictedEffectiveAccessScope(), nil
 	}
-	clusterIDs := make([]string, 0, len(c[0]))
-	for clusterID := range c.topLevelKeys() {
-		clusterIDs = append(clusterIDs, clusterID.String())
+
+	// Drill down to AccessMode level
+	if !c.accessKeys.Contains(AccessModeScopeKey(resource.Access)) {
+		return effectiveaccessscope.DenyAllEffectiveAccessScope(), nil
+	}
+	if c.resourceKeys.Cardinality() == 0 &&
+		c.clusterKeys.Cardinality() == 0 &&
+		c.resourceKeys.Cardinality() == 0 {
+		return effectiveaccessscope.UnrestrictedEffectiveAccessScope(), nil
+	}
+
+	// Drill down to Resource level
+	targetResource := resource.Resource.GetResource()
+	targetReplacingResource := resource.Resource.GetReplacingResource()
+	if !c.resourceKeys.Contains(ResourceScopeKey(targetResource)) &&
+		(targetReplacingResource == nil || !c.resourceKeys.Contains(ResourceScopeKey(*targetReplacingResource))) {
+		return effectiveaccessscope.DenyAllEffectiveAccessScope(), nil
 	}
-	if len(c) == 1 {
+	if c.clusterKeys.Cardinality() == 0 &&
+		c.namespaceKeys.Cardinality() == 0 {
+		return effectiveaccessscope.UnrestrictedEffectiveAccessScope(), nil
+	}
+
+	// Cluster and Namespace level
+	clusterIDs := make([]string, 0, c.clusterKeys.Cardinality())
+	for clusterKey := range c.clusterKeys {
+		clusterIDs = append(clusterIDs, clusterKey.String())
+	}
+	if c.namespaceKeys.Cardinality() == 0 {
 		return effectiveaccessscope.FromClustersAndNamespacesMap(clusterIDs, nil), nil
 	}
-	namespaces := make([]string, 0, len(c[1]))
-	for namespace := range c.next().topLevelKeys() {
-		namespaces = append(namespaces, namespace.String())
+	namespaces := make([]string, 0, c.namespaceKeys.Cardinality())
+	for namespaceKey := range c.namespaceKeys {
+		namespaces = append(namespaces, namespaceKey.String())
 	}
-	clusterNamespaceMap := make(map[string][]string, 0)
+	clusterNamespaceMap := make(map[string][]string, len(clusterIDs))
 	for clusterIx := range clusterIDs {
 		clusterID := clusterIDs[clusterIx]
 		clusterNamespaceMap[clusterID] = namespaces
 	}
 	return effectiveaccessscope.FromClustersAndNamespacesMap(nil, clusterNamespaceMap), nil
 }
 
-func (c allowFixedScopesCheckerCore) topLevelKeys() scopeKeySet {
-	return c[0]
-}
-
-func (c allowFixedScopesCheckerCore) next() allowFixedScopesCheckerCore {
-	return c[1:]
+// AllowFixedScopes returns a scope checker core that allows those scopes that
+// are in the cross product of all individual scope key lists. I.e.,
+//
+//	AllowFixedScopes(
+//		AccessModeScopeKeys(storage.Access_READ, storage.Access_READ_WRITE),
+//		ResourceScopeKeys(resources.CLUSTER),
+//	)
+//
+// returns a scope checker core that allows read and write access to all cluster resources.
+func AllowFixedScopes(keyLists ...[]ScopeKey) ScopeCheckerCore {
```

Comment:
nit: I like to read code top to bottom. Here all helper functions are at the top and the public functions are at the bottom. How about splitting this file?